### ISSUE TITLE
NTK-aware Scaled RoPE

### ISSUE THREAD

**Issue by Jingyu-Fan** [lucidrains/x-transformers#171]
Hi @lucidrain, wondering whether you are interested in incorporating the scaled RoPE to enable models to have extended context size without any fine-tuning. 

The first discussion related was [here](https://www.reddit.com/r/LocalLLaMA/comments/14lz7j5/ntkaware_scaled_rope_allows_llama_models_to_have/), where [bloc97](https://www.reddit.com/user/bloc97/) found that by simply changing 3 lines of the RoPE code, the LLaMA 7b model can have 8k+ context size without any fine-tuning and performance loss! Related code can be found at [HF](https://huggingface.co/kaiokendev/superhot-13b-16k-no-rlhf-test/blob/main/llama_rope_scaled_monkey_patch.py) and [colab](https://colab.research.google.com/drive/1VI2nhlyKvd5cw4-zHvAIk00cAVj2lCCC#scrollTo=e431d2cd).

Furthermore, the author of the RoPE paper(Jianlin Su) became interested and posted 2 blogs on this( [blog1](https://kexue.fm/archives/9675) and [blog2](https://kexue.fm/archives/9706), these 2 blogs are in Chinese), where he found the performance can be further improved by: 
1) log n scaling of the attention (see [here](https://kexue.fm/archives/8823), equations 4 and 5);
2) mixed decimal (see [blog2](https://kexue.fm/archives/9706), equation 9)

Curious about your thoughts?

**Comment by lucidrains** [comment]
@Jingyu-Fan hey Jingyu, i believe i have that too as `rotary_interpolation_factor` https://github.com/lucidrains/x-transformers/blob/main/x_transformers/x_transformers.py#L901 you would simply set this to be higher than 1. on a trained model and fine tune on longer context

**Comment by lucidrains** [comment]
so i have been following Jianlin Su's blog off and on. the log n scaling was notably used in Omegafold (a LLM approach to protein folding). however, i gave it a try on some contrived experiments and did not see much an improvement. willing to change my mind if you have experiments to share (and i can replicate)

2. i haven't seen this yet and will read it later today! can you tell me the gist of it?

**Comment by lucidrains** [comment]
for point 1, i want to make sure this is absolutely necessary before adding it, since calculating sequence length for AR models (as well as variable lengthed bidirectional models) introduces complexity

**Comment by lucidrains** [comment]
@Jingyu-Fan oh i misread the original post. w/o fine tuning is news to me, let me take a closer look today

**Comment by lucidrains** [comment]
ohh, this is cool; now i'm wondering if randomizing the base (or interpolation factor) during training can have some regularizing effect, protect against position overfitting

maybe i can quickly add this, as i do like what i see in the reddit post

now how do i cite a reddit post? lol

**Comment by lucidrains** [comment]
ok done

i'll read what improvements Jianlin Su came up with in his blog post later today! have you tried it?

**Comment by lucidrains** [comment]
also, if you know of a better way to cite reddit posts, let me know lol

**Comment by lucidrains** [comment]
@Jingyu-Fan hey, glad to hear! also got your email; i'm interested in the same topics as you are

if you look at the charts in the reddit post, there seems to be limit to the extrapolation. i also went through Jianlin Su's blog posts, but the improvements seem very incremental

i think your best bet is to carry out some fine tuning at this point

**Comment by lucidrains** [comment]
@Jingyu-Fan so this isn't followed up in the literature much, but some of us in the open source community had a lot of luck using dynamic positional bias for length extrapolation. however, in another project, even that has limits at around 4x the sequence length being trained

if you are training models from scratch, you should [give it a try](https://github.com/lucidrains/x-transformers#dynamic-positional-bias)

**Comment by lucidrains** [comment]
@Jingyu-Fan yes indeed the sequences you are working on are very long

i would recommend taking a look at this new work out of stanford https://arxiv.org/abs/2306.15794 tldr: there is room to capture long distance interactions with subquadratic methods. supplemented with some sparse flash attention, it may go a long way

**Comment by lucidrains** [comment]
> @lucidrains Thanks for the suggestions!! I tried dynamic positional bias before and indeed saw some improvements. Since I expect over 10x the training length, incorporating longer context in the model training seems to be more hopeful...
> 
> For the continuation of the discussion, Jianlin Su just posted a new blog on this today([link](https://kexue.fm/archives/9708)). By setting a window size _w_ and any relative distances in RoPE larger than _w_ to _w_, he was able to almost eliminate the performance loss in length extrapolation. However, that seems to come with incompatibility with flash attention due to operations on the attention matrix... His code is on github. ([link](https://github.com/bojone/rerope/tree/main)) I will take a close look later.

yea, i can take a look at rerope later

but i can safely assure you that one probably cannot hope to extrapolate to unbounded lengths without some amount of fine tuning. not the amount that you are looking for, imo

**Comment by lucidrains** [comment]
@Jingyu-Fan also, tell your wife "go blue"  :smile:

**Comment by bojone** [comment]
> so i have been following Jianlin Su's blog off and on. the log n scaling was notably used in Omegafold (a LLM approach to protein folding). however, i gave it a try on some contrived experiments and did not see much an improvement. willing to change my mind if you have experiments to share (and i can replicate)
> 
> 2. i haven't seen this yet and will read it later today! can you tell me the gist of it?

Logn scaling will hardly change the training effect, but it will alter the extrapolation effect of length, both ntk-rope and rerope will have an impact.

**Comment by bojone** [comment]
> but i can safely assure you that one probably cannot hope to extrapolate to unbounded lengths without some amount of fine tuning. not the amount that you are looking for, imo

Based on the current results, I agree with this. But out of anthropomorphic beliefs, I think length extrapolation is a capability that LLM should have, so I have been actively exploring this.

**Comment by lucidrains** [comment]
> 

hey Jianlin! yes, i did evaluate for length extrapolation, up to 4x the length. i will try again with all your new findings soon when i set aside some time

**Comment by lucidrains** [comment]
i really need to check in my experimental code for length extrapolation, so i can quickly evaluate this

set aside tomorrow morning to give rerope a try

**Comment by lucidrains** [comment]
reopening as a reminder

**Comment by bojone** [comment]
> i really need to check in my experimental code for length extrapolation, so i can quickly evaluate this
> 
> set aside tomorrow morning to give rerope a try

My test data and scripts have all been synced to https://github.com/bojone/rerope/. You're welcome to refer to them, and I look forward to your guidance.

**Comment by lucidrains** [comment]
@bojone should rerope also work for bidirectional attention?

**Comment by lucidrains** [comment]
@bojone i see

does the technique rely on a scale that is a function of sequence length? flash attention is not designed to take in different scales per row

do you still see the effects even without the log(L) scaling?

**Comment by bojone** [comment]
@lucidrains 
In CausalLM, logn scale can be directly multiplied to the query, which does not affect our use of flash attention. 
-Reference: https://github.com/bojone/rerope/blob/main/ntk_patch.py#L22

However, unfortunately, ReRoPE is currently not compatible with flash attention. But this is a problem with ReRoPE itself and has nothing to do with logn scale.

**Comment by lucidrains** [comment]
@bojone oh yes, you are right that logn scaling can be applied on queries :pray: forgot about that

regardless, i should check in a simple length extrapolation experiment script, as it has become a hot research topic

**Comment by lucidrains** [comment]
@bojone congratulations on rotary embeddings btw :smile: 

what a success the technique has become

**Comment by lucidrains** [comment]
<img width="980" alt="Screen Shot 2023-08-24 at 9 35 44 AM" src="https://github.com/lucidrains/x-transformers/assets/108653/4f4d7327-083f-48a3-8647-cad331def32b">

haha, this trick got utilized in today's open sourced code llama, and got a nod from the Meta researchers!

**Comment by lucidrains** [comment]
somebody tell `/u/bloc97`!

**Comment by lucidrains** [comment]
@Jingyu-Fan btw, [someone](https://github.com/lucidrains/enformer-pytorch/issues/30) over at the `enformer-pytorch` repo is also interested in what you are working on. maybe room to collaborate

**Comment by lucidrains** [comment]
i'm going to close this, as it is done
